File: Design Pattern/Easy/stack_using_two_queue.py
#User function Template for python3
'''
    :param x: value to be inserted
    :return: None

    queue_1 = [] # first queue
    queue_2 = [] # second queue
   '''
# https://practice.geeksforgeeks.org/problems/stack-using-two-queues/1#
# Implement a Stack using two queues q1 and q2.
#Function to push an element into stack using two queues.
def push(x):
    
    # global declaration
    global queue_1
    global queue_2
    
    if len(queue_1) == 0:
        queue_1.append(x)
        return
    
    while(len(queue_1)):
        queue_2.append(queue_1[0])
        queue_1.pop(0)
        
    queue_1.append(x)
    
    while(len(queue_2)):
        
        queue_2.pop(0)


#Function to pop an element from stack using two queues.     
def pop():
    
    # global declaration
    global queue_1
    global queue_2
    
    if (len(queue_1) == 0):
        return -1
        
    val = queue_1[0]
    queue_1.pop(0)
    
    return val

#{ 
#  Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys

#Contributed by : Nagendra Jha

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())

queue_1 = [] # first queue
queue_2 = [] # second queue

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        n = int(input())
        a = list(map(int,input().strip().split()))
        i = 0
        while i<len(a):
            if a[i] == 1:
                push(a[i+1])
                i+=1
            else:
                print(pop(),end=" ")
            i+=1

        # clear both the queues
        queue_1 = []
        queue_2 = []
        print()
# } Driver Code Ends

File: Design Pattern/Easy/stack_using_two_queue.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

class QueueStack{
private:
    queue<int> q1;
    queue<int> q2;
public:
    void push(int);
    int pop();
};


int main()
{
    int T;
    cin>>T;
    while(T--)
    {
        QueueStack *qs = new QueueStack();

        int Q;
        cin>>Q;
        while(Q--){
        int QueryType=0;
        cin>>QueryType;
        if(QueryType==1)
        {
            int a;
            cin>>a;
            qs->push(a);
        }else if(QueryType==2){
            cout<<qs->pop()<<" ";

        }
        }
        cout<<endl;
    }
}
// } Driver Code Ends


/* The structure of the class is
class QueueStack{
private:
    queue<int> q1;
    queue<int> q2;
public:
    void push(int);
    int pop();
};
 */
// https://practice.geeksforgeeks.org/problems/stack-using-two-queues/1#
// Implement a Stack using two queues q1 and q2.
//Function to push an element into stack using two queues.
void QueueStack :: push(int x)
{
        if(q1.empty()) q1.push(x);
        else
        {
            while(!q1.empty())
            {
                int top = q1.front();
                q1.pop();
                q2.push(top);
            }
            
            q1.push(x);
            while(!q2.empty())
            {
                int top = q2.front();
                q2.pop();
                q1.push(top);
            }
        }
}

//Function to pop an element from stack using two queues. 
int QueueStack :: pop()
{
        if(q1.empty()) return -1;
        int val = q1.front();
        q1.pop();
        return val;
}
